etna/metrics/base.py

```diff
@@ -227,17 +236,18 @@ def __call__(self, y_true: TSDataset, y_pred: TSDataset) -> Union[float, Dict[st
         """
         self._log_start()
         self._validate_segment_columns(y_true=y_true, y_pred=y_pred)
+        self._validate_timestamp_columns(y_true=y_true, y_pred=y_pred)
+
+        df_true = y_true[:, :, "target"].sort_index(axis=1)
+        df_pred = y_pred[:, :, "target"].sort_index(axis=1)
 
-        segments = set(y_true.df.columns.get_level_values("segment"))
         metrics_per_segment = {}
-        for segment in segments:
-            self._validate_timestamp_columns(
-                timestamp_true=y_true[:, segment, "target"].dropna().index,
-                timestamp_pred=y_pred[:, segment, "target"].dropna().index,
-            )
-            metrics_per_segment[segment] = self.metric_fn(
-                y_true=y_true[:, segment, "target"].values, y_pred=y_pred[:, segment, "target"].values, **self.kwargs
-            )
+        segments = df_true.columns.get_level_values("segment").unique()
```

Comment:
Shouldn't it be sorted as index in the dataframe is sorted?

Reply:
May be we need test for such behaviour(input datasets have unsorted segments)

Reply:
Ok, I'll try to add a test on this.